comp/core/secrets/secretsimpl/secrets.go

```diff
@@ -199,17 +210,61 @@ func isEnc(str string) (bool, string) {
 // Resolve replaces all encrypted secrets in data by executing "secret_backend_command" once if all secrets aren't
 // present in the cache.
 func (r *secretResolver) Resolve(data []byte, origin string) ([]byte, error) {
-	return r.resolve(data, origin, nil)
+	return r.resolve(data, origin, nil, nil)
 }
 
-// ResolveWithCallback resolves the secrets in the given yaml data calling the callback with the YAML path of
+// RegisterResolveCallback resolves the secrets in the given yaml data calling the callback with the YAML path of
 // the secret handle and its value
-func (r *secretResolver) ResolveWithCallback(data []byte, origin string, cb secrets.ResolveCallback) error {
-	_, err := r.resolve(data, origin, cb)
+func (r *secretResolver) RegisterResolveCallback(data []byte, origin string, cb secrets.ResolveCallback) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	// enable refreshing data by storing required parameters
+	r.refreshData = data
+	r.refreshOrigin = origin
+	r.refreshCallback = cb
+	if r.refreshInterval > 0 && r.lastRefresh.IsZero() {
```

Comment:
It looks like that is only the first time we've called `RegisterResolveCallback` that this gorouotine will be created, is that expected? Is `RegisterResolveCallback` meant to be called once? What would happen is someone calls this function twice with different `secrets.ResolveCallback`?

Reply:
Yes, that is expected. The secretResolver is called numerous times with `.Resolve()`, but only once with `.RegisterResolveCallback()`, in `pkg/config/config.go` passing in the config textual data and notification callback. So in the current design, there is a one-to-one connection between that config data and the callback. We may revisit this in the future, so that individual components register to receive the callback, but that would require singling out the config data as being the text over which to resolve secrets. I'd rather wait for that use case to arrive than to anticipate it early. I added a check to this function to ensure its only called once per secrets object.

Reply:
I think the secret comp should be agnostic of its use. Having a method that can't be called more that once is error-prone in a context of a component.

Also we might want to resolve secrets in other configuration this way. For example, a single process might want to resolve secrets from both `datadog.yaml` and `system-probe.yaml`. I think we need a generic and future proof solution to refresh secrets.

I'm not a fan of saving the yaml internally since the data in the configuration could be updated by outside sources (remote-config, CLI, runtime-settings, ...). Since we have the source, context and the yaml path of all secrets in the `cache`, we could directly refresh them and them notify a list of callbacks. We would call the secret executable with all our secrets, refresh the cache and for each update value emit a notification with the source, the yaml path and the new value. Then it's up to the registered caller to update themselves.

That way we can refresh any secrets, not only the ones link to a single configuration and we won't need to parse again the configuration each time we want to check for new values. WDYT ?

Side note: we want to limit this to the API_KEY for now, but we should be ready to support more.